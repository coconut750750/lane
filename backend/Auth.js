import firebase from 'firebase';

import { clientId } from 'lane/config';

export function getUserID() {
    return firebase.auth().currentUser.uid;
}

function isUserEqual(googleUser, firebaseUser) {
    if (firebaseUser) {
        var providerData = firebaseUser.providerData;
        for (var i = 0; i < providerData.length; i++) {
            if (providerData[i].providerId === firebase.auth.GoogleAuthProvider.PROVIDER_ID &&
                providerData[i].uid === googleUser.getBasicProfile().getId()) {
                return true;
            }
        }
    }
    return false;
}

function onSignIn(googleUser) {
    // We need to register an Observer on Firebase Auth to make sure auth is initialized.
    var unsubscribe = firebase.auth().onAuthStateChanged(firebaseUser => {
        unsubscribe();
        // Check if we are already signed-in Firebase with the correct user.
        if (!isUserEqual(googleUser, firebaseUser)) {
            // Build Firebase credential with the Google ID token.
            var credential = firebase.auth.GoogleAuthProvider.credential(
                googleUser.idToken,
                googleUser.accessToken
             );
            // Sign in with credential from the Google user.
            firebase
                .auth()
                .signInWithCredential(credential)
                .then(function(result) {
                    if (result.additionalUserInfo.isNewUser) {
                        firebase.database()
                            .ref('users')
                            .child(result.user.uid)
                            .set({
                                email: result.user.email,
                                profile: result.additionalUserInfo.profile.picture,
                                firstname: result.additionalUserInfo.profile.given_name,
                                lastname: result.additionalUserInfo.profile.family_name
                            });
                        firebase.database()
                            .ref('emails')
                            .child(result.user.uid)
                            .set(result.user.email);
                    }
                })
                .catch(function(error) {
                    // Handle Errors here.
                    var errorCode = error.code;
                    var errorMessage = error.message;
                    // The email of the user's account used.
                    var email = error.email;
                    // The firebase.auth.AuthCredential type that was used.
                    var credential = error.credential;
                });
        } else {
            console.log('User already signed-in Firebase.');
        }
    });
}

export async function signInWithGoogleAsync() {
    try {
        const result = await Expo.Google.logInAsync({
            behavior: 'web',
            androidClientId: clientId.android,
            iosClientId: clientId.ios,
            scopes: ['profile', 'email']
        });

        if (result.type === 'success') {
            onSignIn(result);
            return result.accessToken;
        } else {
            return {canceled: true};
        }
    } catch (e) {
        return {error: true};
    }
}

export function signOut() {
    firebase.auth().signOut();
}

export function checkIfLoggedIn(checkComplete) {
    firebase.auth().onAuthStateChanged(user => {
        checkComplete(user != undefined);
    });
}

export async function getIdFromEmail(email, onComplete) {
    firebase.database().ref('emails').once('value', snapshot => {
        var emails = snapshot.val();
        for (var id in emails) {
            if (emails[id] === email) {
                onComplete(id);
                return;
            }
        }
        onComplete(undefined);
    });
}