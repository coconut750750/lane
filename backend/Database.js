import firebase from 'firebase';
import { convertTimestampToDateString } from 'lane/utils/utils';

let USERS = 'users';
let LANES = 'lanes';
let SHARED = 'shared';

export async function addLaneToUser(userId, laneId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .child(LANES).child(laneId)
        .set(0);
}

export async function removeLaneFromUser(userId, laneId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .child(LANES).child(laneId)
        .remove();
}

export async function removeUser(userId) {
    await firebase.database()
        .ref(USERS).child(userId)
        .remove();
}

export async function addUserToLane(userId, laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .child(SHARED).child(userId)
        .set(0);
}

export async function removeUserFromLane(userId, laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .child(SHARED).child(userId)
        .remove();
}

export async function removeAllSharedUsers(laneObj) {
    if (laneObj.shared != null) {
        await Promise.all(
            Object.keys(laneObj.shared).map( id => 
                removeLaneFromUser(id, laneObj.id)
            )
        );
    }
}

export async function removeLane(laneId) {
    await firebase.database()
        .ref(LANES).child(laneId)
        .remove();
}

export async function addPhoto(blob, photo, laneId) {
    let photoId = photo.md5;

    const snapshot = await firebase.storage()
        .ref()
        .child(laneId)
        .child(photoId)
        .put(blob);
    blob.close();

    const photoUrl = await snapshot.ref.getDownloadURL();

    await firebase.database()
        .ref(LANES).child(laneId)
        .child('photos').child(photoId)
        .set({
            uri: photoUrl,
            width: photo.image.width,
            height: photo.image.height,
            timestamp: convertTimestampToDateString(photo.timestamp),
        });
}

export async function removePhoto(photoId, laneId) {
    await firebase.storage()
        .ref()
        .child(laneId)
        .child(photoId)
        .delete();

    await firebase.database()
        .ref(LANES).child(laneId)
        .child('photos').child(photoId)
        .remove();
}

export async function removeAllLanePhotos(laneObj) {
    await Promise.all(
        Object.keys(laneObj.photos).map( photoId => 
            firebase.storage().ref().child(laneObj.id).child(photoId).delete()
        )
    );
}

export async function pushLane(ownerId, laneObj) {
    laneObj.owner = ownerId;
    return await firebase.database().ref(LANES).push(laneObj).key;
}

export async function onLaneUpdate(processLanes) {
    var userid = firebase.auth().currentUser.uid;
    firebase.database().ref(USERS).child(userid).child(LANES).on('value', datasnapshot => {
        var lanes = datasnapshot.val();
        var laneObjs = {};

        if (!lanes) {
            processLanes(laneObjs);
        } else {
            var markings = Object.keys(lanes).map( laneId =>
                firebase.database().ref(LANES).child(laneId).once('value', lanesnapshot => {
                    var lane = lanesnapshot.val();
                    laneObjs[laneId] = {...lane, id: laneId};
                })
            );
            
            Promise.all(markings).then(lanes => {
                processLanes(laneObjs);
            });
        }
    });
}

export async function uploadImageAsync(laneId, photo) {
    const blob = await new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.onload = function() {
            resolve(xhr.response);
        };
        xhr.onerror = function(e) {
            console.log(e);
            reject(new TypeError('Network request failed'));
        };
        xhr.responseType = 'blob';
        xhr.open('GET', photo.image.uri, true);
        xhr.send(null);
    });

    await addPhoto(blob, photo, laneId);
}

export async function shareLane(laneId, otherUserId) {
    await Promise.all([
        addLaneToUser(otherUserId, laneId),
        addUserToLane(otherUserId, laneId)
    ]);
}

export function deleteLane(laneId, onComplete) {
    firebase.database().ref(LANES).child(laneId).once('value', lanesnapshot => {
        var laneObj = lanesnapshot.val();
        laneObj.id = laneId;
        Promise.all([
            removeAllSharedUsers(laneObj),
            removeLaneFromUser(laneObj.owner, laneId),
            removeAllLanePhotos(laneObj),
            removeLane(laneId)
        ]).then(_ => {
            onComplete();
        });
    });
}