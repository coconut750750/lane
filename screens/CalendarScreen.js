import React, { Component } from 'react';
import {
    View,
    StyleSheet,
    ActivityIndicator,
    Modal,
    Animated,
    TouchableWithoutFeedback,
} from 'react-native';
import { FAB, Text, IconButton, Snackbar } from 'react-native-paper';
var _ = require('lodash');

import { getUserID, signOut, getIdFromEmail } from 'lane/backend/Auth';
import { onLaneUpdate, shareLane, deleteLane, unsubscribeLane } from 'lane/backend/Database';

import CalendarHeader from 'lane/components/CalendarHeader';
import LaneCalendar from 'lane/components/LaneCalendar';
import LaneContent from 'lane/components/LaneContent';

import Colors from 'lane/constants/Colors'
import Layout from 'lane/constants/Layout'

import schedulePeriods from 'lane/utils/PeriodScheduling';
import { constructPeriodFromLane, propagatePeriod, setupScheduledMarkings, getValidLanes } from 'lane/utils/PeriodTools';
import { months } from 'lane/utils/TimeTools';

export default class CalendarScreen extends Component {
    constructor(props) {
        super(props);

        this.lanes = {};
        this.periods = [];
        this.selectedLanes = [];
        this.state = {
            markings: {},
            currentLane: 0,
            loading: true,

            // animations
            scrollAnim: new Animated.Value(0),

            // snackbar
            snackVisible: false,
            snackMessage: '',
        };
    }

    componentDidMount() {
        onLaneUpdate(this.processLane.bind(this), this.unprocessLane.bind(this));
    }

    alert(message) {
        this.setState({
            snackVisible: true,
            snackMessage: message
        });
    }

    processLane(lane) {
        if (lane === undefined) {
            this.setState({ loading: false });
            return;
        }

        this.lanes[lane.id] = lane;
        this.processAll();
    }

    unprocessLane(laneId) {
        delete this.lanes[laneId];
        this.processAll();
    }

    processPeriods(periods) {
        var scheduled = schedulePeriods(periods);
        return setupScheduledMarkings(scheduled);
    }

    processAll() {
        this.periods = [];
        var propagatedPeriods = [];
        _.forEach(this.lanes, (laneObj, id) => {
            const period = constructPeriodFromLane(laneObj)
            this.periods.push(period);
            propagatedPeriods = propagatedPeriods.concat(propagatePeriod(period));
        });
        this.periods = schedulePeriods(this.periods);
        const markings = this.processPeriods(propagatedPeriods);

        this.setState({
            markings: markings,
            loading: false,
        });
        this.unselect();
    }

    markSelectedLane() {
        const laneId = this.selectedLanes[this.state.currentLane];
        var laneObj = this.lanes[laneId];
        if (laneObj === undefined) {
            return {};
        }
        var period = constructPeriodFromLane(laneObj);
        period.height = 0;
        return setupScheduledMarkings([period]);
    }

    select(lanes) {
        this.selectedLanes = lanes;
        this.setState({
            currentLane: 0,
            scrollAnim: new Animated.Value(0),
        });
    }

    unselect() {
        this.selectedLanes = []
        this.setState({ 
            selectedDay: '',
            currentLane: 0,
            scrollAnim: new Animated.Value(0),
        });
    }

    onPressDay(date) {
        if (this.state.selectedDay != date) {
            this.setState({ selectedDay: date }, async () => {
                await new Promise(resolve => setTimeout(resolve, 0));
                this.getLanes(date)
            });
        } else {
            this.unselect();
        }
    }

    getLanes(date) {
        const selectedLanes = getValidLanes(this.periods, date);
        if (!_.isEqual(selectedLanes, this.selectedLanes)) {
            this.select(selectedLanes);
        }
    }

    onUpdateMonth(month) {
        monthStr = `${months[month.month - 1]} ${month.year}`
        if (this.state.month != monthStr) {
            this.setState({
                month: monthStr,
            });
        }
    }

    onBackLane() {
        this.setState({
            currentLane: (this.state.currentLane - 1) % this.selectedLanes.length
        });
    }

    onNextLane() {
        this.setState({
            currentLane: (this.state.currentLane + 1) % this.selectedLanes.length
        });
    }

    onEditLane(laneObj) {
        this.props.navigation.navigate('Edit', { laneObj: laneObj });
    }

    onShareLane(laneObj, email) {
        this.setState({
            loading: true,
        });
        getIdFromEmail(email, id => {
            if (id === undefined) {
                this.alert('Email doesn\'t exist!');
            } else {
                shareLane(laneObj.id, id);
            }
            this.setState({ loading: false });
        });
    }

    onDeleteLane(laneObj) {
        this.unselect();
        this.setState({
            loading: true,
        });
        deleteLane(laneObj.id, () => {
            this.setState({
                loading: false,
            });
        });
    }

    onUnsubscribeLane(laneObj) {
        this.unselect();
        unsubscribeLane(getUserID(), laneObj.id);
    }

    renderLoading() {
        return (
            <View style={styles.container}>
                <ActivityIndicator 
                    size="large"
                    color={Colors.primary} />
            </View>
        );
    }

    renderLaneContent(transform) {
        if (this.selectedLanes.length > 0) {
            const laneId = this.selectedLanes[this.state.currentLane];
            return (
                <LaneContent
                    lane={ this.lanes[laneId] }
                    onBackLane={ () => this.onBackLane() }
                    onNextLane={ () => this.onNextLane() }
                    onEditLane={ lane => this.onEditLane(lane) }
                    onShareLane={ (lane, email) => this.onShareLane(lane, email) }
                    onDeleteLane={ lane => this.onDeleteLane(lane) }
                    onUnsubscribeLane={ lane => this.onUnsubscribeLane(lane) }
                    calendarHeight={ Layout.calendarHeight }
                    headerTransform={ transform }
                    onScroll={Animated.event(
                        [{ nativeEvent: { contentOffset: { y: this.state.scrollAnim } } }],
                        { useNativeDriver: true },
                    )}
                />
            );
        }

        return null;
    }

    renderLaneCalendar(transform) {
        var markings = { ...this.state.markings };
        var height = Layout.window.height;

        if (this.selectedLanes.length > 0) {
            markings = this.markSelectedLane();
            height = Layout.calendarHeight;
        }

        return (
            <Animated.View style={{ ...styles.calendar, transform: transform }}>
                <LaneCalendar
                    markings={ markings }
                    onDayPress={ date => this.onPressDay(date) }
                    onMonthChange={ month => this.onUpdateMonth(month[0]) }
                    selectedDay={ this.state.selectedDay }
                    height={ height }
                />
                <CalendarHeader
                    month={this.state.month}
                    openDrawer={ () => this.props.navigation.openDrawer() }
                    signOut={ () => signOut() }/>
            </Animated.View>
        );
    }

    render() {
        if (this.state.loading) {
            return this.renderLoading();
        }

        const calendarTranslate = this.state.scrollAnim.interpolate({
            inputRange: [0, Layout.calendarHeight],
            outputRange: [0, -Layout.calendarHeight],
            extrapolate: 'clamp',
        });
        const transform = [{ translateY: calendarTranslate }];

        // content rendered first because it needs to be rendered below (z-dimension) the calendar
        return (
            <View style={styles.container}>
                {this.renderLaneContent(transform)}

                {this.renderLaneCalendar(transform)}

                

                <FAB
                    style={styles.fab}
                    icon="add"
                    onPress={ () => this.props.navigation.navigate('Create') }
                />
                <Snackbar
                    visible={this.state.snackVisible}
                    duration={3000}
                    onDismiss={() => this.setState({ snackVisible: false, snackMessage: '' })}>
                    { this.state.snackMessage }
                </Snackbar>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center'
    },
    fab: {
        position: 'absolute',
        margin: 16,
        right: 0,
        bottom: 0,
    },
    calendar: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
    },
})