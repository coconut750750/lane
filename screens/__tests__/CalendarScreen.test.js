import React from 'react'
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import * as auth from 'lane/backend/Auth';
import * as db from 'lane/backend/Database';

import CalendarScreen from '../CalendarScreen'

jest.mock("lane/backend/Auth", () => ({
    getIdFromEmail: jest.fn(),
}));
jest.mock("lane/backend/Database", () => ({
    onLaneUpdate: jest.fn(),
    shareLane: jest.fn(),
    deleteLane: jest.fn(),
}));

const mockNavigate = jest.fn( dest => {} );
const navigation = { navigate: mockNavigate };

Enzyme.configure({ adapter: new Adapter() });

describe('CalendarScreen', () => {
    const singleDayEmptyLane = {
        id: 'testing-lane0',
        color: '#ffffff',
        endDate: '2000-01-01',
        owner: 'testing-owner0',
        photos: {},
        startDate: '2000-01-01',
        title: 'testing-title0'
    };

    const singleDayLane = {
        id: 'testing-lane1',
        color: '#ffffff',
        endDate: '2000-01-05',
        owner: 'testing-owner1',
        photos: {id0 : {uri: 'uri', width: 100, height: 100}},
        startDate: '2000-01-01',
        title: 'testing-title1'
    }

    const setup = () => {
        const component = shallow(<CalendarScreen navigation={ navigation }/>);
        const instance = component.instance();
        return { component, instance };
    }

    const setupWithLanes = (lanes) => {
        db.onLaneUpdate.mockImplementation( process => process(lanes) );
        return setup();
    }

    beforeEach(() => {
        auth.getIdFromEmail.mockReset();
        db.onLaneUpdate.mockReset();
        db.shareLane.mockReset();
        db.deleteLane.mockReset();
    });

    it('immediately renders loading screen', () => {
        const { component, instance } = setup();
        const screen = renderer.create(instance.render()).toJSON();
        const loading = renderer.create(instance.renderLoading()).toJSON();

        expect(instance.state.loading).toBeTruthy();
        expect(screen).toEqual(loading);
        expect(screen).toMatchSnapshot();
    });

    it('did start lane listender', () => {
        const { component, instance } = setup();

        expect(db.onLaneUpdate).toHaveBeenCalled();
    });

    it('renders zero lanes correctly', () => {
        const { component, instance } = setupWithLanes([]);

        expect(Object.keys(instance.state.lanes).length).toBe(0);
        expect(Object.keys(instance.state.markings).length).toBe(0);
        expect(instance.state.loading).toBeFalsy();

        const screen = renderer.create(instance.render()).toJSON();
        expect(screen).toMatchSnapshot();
    });

    it('processes a single-day empty lane correctly', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        expect(Object.keys(instance.state.lanes).length).toBe(1);
        expect(Object.keys(instance.state.markings).length).toBe(1);
        expect(instance.state.loading).toBeFalsy();
    });

    it('selects a single-day empty lane correctly', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);
        instance.getLanes(singleDayEmptyLane.startDate);

        expect(instance.state.selectedLanes).toEqual([0]);
    });

    it('processes two lanes with 5-day span correctly', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane, singleDayLane]);

        expect(Object.keys(instance.state.lanes).length).toBe(2);
        expect(Object.keys(instance.state.markings).length).toBe(5);
        expect(instance.state.loading).toBeFalsy();

    });

    it('selects two lanes correctly', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane, singleDayLane]);

        instance.getLanes(singleDayEmptyLane.startDate);
        expect(instance.state.selectedLanes).toEqual([0, 1]);

        instance.getLanes(singleDayLane.endDate);
        expect(instance.state.selectedLanes).toEqual([1]);
    });

    it('unselects when lanes are updated', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        expect(instance.state.selectedLanes).toEqual([]);

        instance.getLanes(singleDayEmptyLane.startDate);
        expect(instance.state.selectedLanes).toEqual([0]);

        instance.processLanes([singleDayEmptyLane, singleDayLane]);
        expect(instance.state.selectedLanes).toEqual([]);
    });

    it('renders share modal', () => {
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        expect(instance.state.shareModalOpen).toBeFalsy();

        instance.onShareLane(singleDayEmptyLane);
        expect(instance.state.shareModalOpen).toBeTruthy();
        expect(instance.state.sharingId).toEqual(singleDayEmptyLane.id);
    });

    it('sends a share invitation to vaild email', () => {
        const id = 'test-id';
        const email = 'test@me.com';

        auth.getIdFromEmail.mockImplementation( (email, callback) => callback(id) );
        db.shareLane.mockImplementation( (laneId, otherUserId) => {} );
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        instance.onShareLane(singleDayEmptyLane);
        instance.onSendShare(email);

        expect(instance.state.shareModalOpen).toBeFalsy();
        expect(db.shareLane).toHaveBeenCalledWith(singleDayEmptyLane.id, id);
    });

    it('sends a share invitation to invaild email', () => {
        const email = 'test@me.com';

        auth.getIdFromEmail.mockImplementation( (email, callback) => callback(undefined) );
        db.shareLane.mockImplementation( (laneId, otherUserId) => {} );
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        instance.onShareLane(singleDayEmptyLane);
        instance.onSendShare(email);

        expect(instance.state.shareModalOpen).toBeFalsy();
        expect(db.shareLane).not.toHaveBeenCalled();
    });

    it('successfully deletes a lane', () => {
        db.deleteLane.mockImplementation( (laneId, callback) => callback() );
        const { component, instance } = setupWithLanes([singleDayEmptyLane]);

        instance.onDeleteLane(singleDayEmptyLane);

        expect(db.deleteLane).toHaveBeenCalled();
    });
});