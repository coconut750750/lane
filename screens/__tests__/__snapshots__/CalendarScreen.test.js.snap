// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarScreen immediately renders loading screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#6200ee"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`CalendarScreen renders zero lanes correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffffbb",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "elevation": 4,
            "justifyContent": "center",
            "margin": 12,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 4,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "rgb(231, 231, 231)",
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgba(0, 0, 0, 0.26)",
                  "bottom": 0,
                  "height": 2,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "transform": Array [
                    Object {
                      "scaleY": 0.5,
                    },
                  ],
                }
              }
            />
            <View
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "opacity": 1,
                  },
                ]
              }
            >
              <Text
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "Helvetica Neue",
                      "textAlign": "left",
                      "writingDirection": "ltr",
                    },
                    Object {
                      "color": "#6200ee",
                      "fontFamily": "Helvetica Neue",
                      "fontSize": 16,
                      "left": 0,
                      "opacity": 0,
                      "paddingHorizontal": 12,
                      "position": "absolute",
                      "top": 19,
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "scale": 1,
                        },
                        Object {
                          "translateX": 0,
                        },
                      ],
                    },
                  ]
                }
              >
                Add email
              </Text>
              <Text
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "Helvetica Neue",
                      "textAlign": "left",
                      "writingDirection": "ltr",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.54)",
                      "fontFamily": "Helvetica Neue",
                      "fontSize": 16,
                      "left": 0,
                      "opacity": 1,
                      "paddingHorizontal": 12,
                      "position": "absolute",
                      "top": 19,
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                        Object {
                          "translateY": 0,
                        },
                        Object {
                          "scale": 1,
                        },
                        Object {
                          "translateX": 0,
                        },
                      ],
                    },
                  ]
                }
              >
                Add email
              </Text>
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder=""
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "fontSize": 16,
                    "margin": 0,
                    "minHeight": 58,
                    "paddingHorizontal": 12,
                    "textAlign": "left",
                    "zIndex": 1,
                  },
                  Object {
                    "paddingBottom": 6,
                    "paddingTop": 24,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "Helvetica Neue",
                    "textAlignVertical": "center",
                  },
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "flex": 0.6,
              }
            }
          />
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "borderRadius": 4,
                "borderStyle": "solid",
                "borderWidth": 0,
                "elevation": 0,
                "flex": 0.2,
                "minWidth": 64,
                "shadowColor": "#000000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.24,
                "shadowRadius": 0,
              }
            }
          >
            <View
              accessibilityRole="button"
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 4,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "Helvetica Neue",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "letterSpacing": 1,
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        undefined,
                        Object {
                          "color": "#6200ee",
                          "fontFamily": "HelveticaNeue-Medium",
                        },
                        Object {
                          "fontFamily": "HelveticaNeue-Medium",
                        },
                      ],
                    ]
                  }
                >
                  CANCEL
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "borderRadius": 4,
                "borderStyle": "solid",
                "borderWidth": 0,
                "elevation": 0,
                "flex": 0.2,
                "minWidth": 64,
                "shadowColor": "#000000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.24,
                "shadowRadius": 0,
              }
            }
          >
            <View
              accessibilityRole="button"
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 4,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "Helvetica Neue",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "letterSpacing": 1,
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        undefined,
                        Object {
                          "color": "#6200ee",
                          "fontFamily": "HelveticaNeue-Medium",
                        },
                        Object {
                          "fontFamily": "HelveticaNeue-Medium",
                        },
                      ],
                    ]
                  }
                >
                  SEND
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <View
    style={
      Object {
        "height": 360,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  />
  <View
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#03dac4",
        "borderRadius": 28,
        "bottom": 0,
        "elevation": 6,
        "margin": 16,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 6,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 28,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "height": 56,
              "width": 56,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#03dac4",
        "borderRadius": 28,
        "bottom": 0,
        "elevation": 6,
        "left": 0,
        "margin": 16,
        "opacity": 1,
        "position": "absolute",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 6,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 28,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "height": 56,
              "width": 56,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 18,
            "height": 36,
            "justifyContent": "center",
            "margin": 6,
            "overflow": "visible",
            "width": 36,
          },
          undefined,
          Object {
            "left": 0,
            "margin": 16,
            "position": "absolute",
            "top": 0,
          },
        ],
      ]
    }
  >
    <View
      style={null}
    >
      <Text />
    </View>
  </View>
</View>
`;
